lib/connections/limiter.go

```diff
@@ -162,3 +263,37 @@ func (b *atomicBool) set(v bool) {
 func (b *atomicBool) get() bool {
 	return atomic.LoadInt32((*int32)(b)) != 0
 }
+
+// Utility functions for atomic operations on device limiters map
+func (lim *limiter) getOrCreateDeviceWriteLimiter(deviceID protocol.DeviceID) (*rate.Limiter, bool) {
+	lim.deviceMapMutex.Lock()
+	defer lim.deviceMapMutex.Unlock()
+	limiter, ok := lim.deviceWriteLimiters[deviceID]
+
+	if !ok {
+		limiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+		lim.deviceWriteLimiters[deviceID] = limiter
+	}
+
+	return limiter, ok
+}
+
+func (lim *limiter) getOrCreateDeviceReadLimiter(deviceID protocol.DeviceID) (*rate.Limiter, bool) {
```

Comment:
The name is terrible.
Just getReadLimiter() or even readLimiter(), and the fact that it creates it underneath is an irrelevant detail that doesn't need to be exposed as part of the function name.